Refactors Toolhouse.Monitoring.Metrics.InstrumentApiCall

bgradin opened the pull request:
Link to gist: https://gist.github.com/bgradin/b6411213cc1a49b06b1236c84e2497ba

This PR can also be used to get the ball rolling for updating the project's major version.

bgradin committed:
Refactors Toolhouse.Monitoring.Metrics.InstrumentApiCall into a separate class, Toolhouse.Monitoring.HttpRequestMetrics

bgradin commented:
## Usage

```cs
IRestResponse<T> Foo(Method method)
{
    var request = new RestRequest("", method);
    return m_client.Execute<T>(request);
}
```

Becomes:

```cs
IRestResponse<T> Foo(Method method)
{
    var request = new RestRequest("", method);
    return HttpRequestMetrics.Instrument("connect", () => m_client.Execute<T>(request));
 }
```

Custom logging for failures, status codes or additional data is also available:

```cs
IRestResponse<T> Foo(Method method)
{
    var request = RestRequest("", method);
    return HttpRequestMetrics.Report("connect", labels =>
    {
        IRestResponse<T> response = m_client.Execute<T>(request);
        labels.AddSuccess();
        labels["action"] = "login";

        if (response.StatusCode != HttpStatusCode.OK)
        {
            labels.AddFailure();
            labels.AddStatusCode(response.StatusCode);
        }

        return response;
    });
}
```

If you don't need a pass-through value, the API can also return `void`:

```cs
void Foo(Method method)
{
    var request = new RestRequest("", method);
    HttpRequestMetrics.Report("connect", () => { m_client.Execute<T>(request); });
}
```

matthinz commented on Toolhouse.Monitoring/Metrics.cs:
```diff
@@ -157,77 +155,6 @@ public static void ObserveHttpRequestDuration(TimeSpan duration)
-        /// </summary>
-        /// <param name="name">Descriptive name of thing being done. Should be a singular noun, e.g. "salesforce".</param>
-        /// <param name="callback">Code to run. Should return true for success, false for failure.</param>
-        public static void InstrumentApiCall(string name, Func<bool> makeRequest)
```
We should leave this method in place, but rework it to use the new instrumentation code under the hood.

matthinz commented on Toolhouse.Monitoring/HttpRequestMetrics.cs:
```diff
@@ -0,0 +1,148 @@
+                .Labels(m_labels.Values).Observe(m_stopwatch.Elapsed.TotalSeconds);
+        }
+
+        readonly Labels m_labels;
```
I'd like the layout of new classes to match layout of existing classes in this project. Please:

- Move fields to the top of classes
- Omit the `m_` prefix
- Ensure all fields have an appropriate access modifier (`private`, `protected`, etc.)

(Ideally we'd have some automated checking of this, but I haven't had the time to set it up here)

bgradin replied:
I agree with the second point as long as there is some way to distinguish private member variables from instance variables and properties. Maybe we should have a more in-depth discussion about this.

matthinz commented on Toolhouse.Monitoring/HttpRequestMetrics.cs:
```diff
@@ -0,0 +1,148 @@
+        /// <param name="name">Descriptive name of thing being done. Should be a singular noun, e.g. "salesforce".</param>
+        /// <param name="request">An action which will perform the request</param>
+        /// <remarks>A mutable Labels instance will be provided to the request.</remarks>
+        public static void Instrument(string name, Action<Labels> request)
```
I feel like:

1. These `Instrument()` calls should be instance methods, not `static`
2. The constructor of this class should be `public` and accept a `name` argument

bgradin replied:
I'd be interested to hear your thoughts on these points - especially the first one. My original idea to have these `static` was due to several things:

1. The original api being `static`
2. Ease of use of a `static` API
3. Not wanting to create an object that only has one data member (the name)

I think maybe a compromise might work well here - we could simply make the existing constructor `public`. That would allow for the pattern shown in the usage note above, but also allow for situations where one request type is executed in multiple places; where it'd be advantageous to share the `name`.

matthinz replied:
Sorry, I'm coming back to this with fresh eyes and I think I've changed my mind a bit 😄.  My concern is mostly around the complexity of the API as a consumer. I think:

- Refactoring all of this into `HttpRequestMetrics` is 💯, but
- The `Metrics` class should be the only place that provides **static** methods for adding metrics to your app.

So I think what I'd like to see is these these various static `Instrument` overloads **moved** to the `Metrics` class (keep the old `InstrumentApiCall` marked as deprecated). That way, the main API for adding metrics is always available via the static `Metrics` class, but you can still `new` up an `HttpRequestMetrics` instance if you want to do something crazy.

matthinz commented on Toolhouse.Monitoring/Labels.cs:
```diff
@@ -0,0 +1,66 @@
+            this[c_successLabelKey] = "0";
+        }
+
+        const string c_backendLabelKey = "backend";
```
Same formatting nitpick applies here; see other classes for example `const` naming

matthinz reviewed (changes requested):
This is looking good, and I appreciate the level of test coverage. Take a crack at the changes I noted and we'll move forward.

bgradin committed:
Moves members to the top of the class and adds access specifier

bgradin committed:
Opens up access to HttpRequestMetrics constructor

This could be useful if the consuming code needs to track instrumentation of a single request type in multiple places

bgradin committed:
Adds back in old method, but marks it as obsolete

matthinz commented on Toolhouse.Monitoring/HttpRequestMetrics.cs:
```diff
@@ -0,0 +1,152 @@
+    /// <summary>
+    /// Provides methods for instrumenting requests to an external API (e.g. via REST / SOAP).
+    /// </summary>
+    public sealed class HttpRequestMetrics
```
Let's not seal this unnecessarily

matthinz commented on Toolhouse.Monitoring/Labels.cs:
```diff
@@ -0,0 +1,66 @@
+
+namespace Toolhouse.Monitoring
+{
+    public sealed class Labels
```
Ditto remove `sealed` here

matthinz commented on Toolhouse.Monitoring/Metrics.cs:
```diff
@@ -161,71 +161,11 @@ public static void ObserveHttpRequestDuration(TimeSpan duration)
         /// <param name="name">Descriptive name of thing being done. Should be a singular noun, e.g. "salesforce".</param>
-        /// <param name="callback">Code to run. Should return true for success, false for failure.</param>
+        /// <param name="makeRequest">Code to run. Should return true for success, false for failure.</param>
+        [Obsolete("Use HttpRequestMetrics.Instrument instead.")]
```
Per above, augment this class with the various `Instrument<T>(...)` overloads and update this `[Obsolete]`

matthinz reviewed (changes requested):
Thank you for poking me about this. I'd like a couple of things moved around to smooth out the external API a bit, then I think we'll be good.

bgradin committed:
Updates Metrics and HttpRequestMetrics based on Matt's feedback

matthinz committed:
Make HttpRequestMetrics know about exceptions

If the code being instrumented throws an exception, note that via a
"success=0" label (no exception results in "success=1"). Then, re-throw
the exception.

matthinz committed:
InstrumentApiCall: keep success-tracking semantics

InstrumentApiCall used return value of the provided Func<bool> to indicate success vs. failure. Here we keep that by instead throwing an exception when the clousure returns false (and catching it to keep it from leaking out).

matthinz committed:
Update version: 1.3.0-alpha

matthinz committed:
Use correct case for reference

This matters on case-sensitive filesystems (i.e. when running tests
under mono on linux).

matthinz committed:
Execute CI tests inside docker container

To get around weirdness in running mono directly in travis:

- Add a Dockerfile that builds projects via mono and runs tests
- Update .travis.yml to build + run Docker container
